sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,858 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.amqp.implementation;
+
+import com.azure.core.amqp.AmqpEndpointState;
+import com.azure.core.amqp.AmqpRetryPolicy;
+import com.azure.core.util.AsyncCloseable;
+import com.azure.core.util.logging.ClientLogger;
+import org.apache.qpid.proton.Proton;
+import org.apache.qpid.proton.message.Message;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Disposable;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxOperator;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Operators;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.concurrent.Queues;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import static com.azure.core.amqp.implementation.ClientConstants.ENTITY_PATH_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.LINK_NAME_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.SUBSCRIBER_ID_KEY;
+
+/**
+ * The Flux operator to stream messages reliably from a messaging entity (e.g., Event Hub, Service Bus Queue)
+ * to downstream subscriber.
+ */
+public final class MessageFlux extends FluxOperator<ReactorReceiver, Message> {
+    private final int prefetch;
+    private final AmqpRetryPolicy retryPolicy;
+
+    /**
+     * Create a message-flux to stream messages from a messaging entity to downstream subscriber.
+     *
+     * @param source the upstream source that, upon request, provide receivers connected to the messaging entity.
+     * @param prefetch the number of messages that the operator should prefetch from the messaging entity (for a
+     *                 less chatty network and faster message processing on the client).
+     * @param retryPolicy the retry policy to use to recover from receiver termination.
+     */
+    protected MessageFlux(Flux<? extends ReactorReceiver> source,
+                          int prefetch,
+                          AmqpRetryPolicy retryPolicy) {
```

Comment:
1. Ctr should be updated to take ConnectionProcessor
2. The drain-loop should be updated to inspect ConnectionProcessor.isDisposed()